src/main/scala/mesosphere/marathon/api/akkahttp/Directives.scala

```diff
@@ -14,14 +18,60 @@ import scala.concurrent.duration._
   */
 object Directives extends AuthDirectives with LeaderDirectives with AkkaDirectives {
 
+  private val marathonApiKeywords = Set("restart", "tasks", "versions")
+
+  //basically it's equivalent to path.takeWhile(s => !marathonApiKeywords.contains(s)) map createPathId
+  private[akkahttp] def findAppId(remainingPath: Path, extractedAppName: Path = Path.Empty): Option[PathId] = {
+    //transforming Path to PathId
+    def generatePathId(p: Path): PathId = {
+      val trimmed = p.toString().reverse.dropWhile(_ == '/').reverse
+      PathId(trimmed)
+    }
+
+    remainingPath match {
+      case Path.Slash(rest) =>
+        findAppId(rest, extractedAppName ++ Path./)
+      case p @ Path.Segment(segment, rest) =>
+        if (marathonApiKeywords.contains(segment)) {
+          Some(generatePathId(extractedAppName))
+        } else {
+          findAppId(rest, extractedAppName + segment)
+        }
+      case Path.Empty =>
+        if (extractedAppName.isEmpty) {
+          None
+        } else {
+          Some(generatePathId(extractedAppName))
+        }
+    }
+  }
+
   /**
-    * Path matcher, that matches a segment only, if it is defined in the given set.
-    * @param set the allowed path segments.
+    * Given the current root group, only extract an existing appId
+    *
+    * This is useful because our v2 API has an unfortunate design decision which leads to ambiguity in our URLs, such as:
+    *
+    *   POST /v2/apps/my-group/restart/restart
+    *
+    * The intention here is to restart the app named "my-group/restart"
+    *
+    * Given the url above, this directive will only extract "my-group/restart" from the path, leaving the rest.
+    *
+    * In case there is no appId exists, it will do best-efforts attempt to find an invalid appId and generate a rejection.
+    *
+    * Inspired by path matching directive in akka-http.
     */
-  class PathIsAvailableInSet(set: Set[String]) extends PathMatcher1[String] {
-    def apply(path: Path) = path match {
-      case Path.Segment(segment, tail) if set(segment) ⇒ Matched(tail, Tuple1(segment))
-      case _ ⇒ Unmatched
+  def extractExistingAppId(rootGroup: RootGroup): Directive1[PathId] = {
```

Comment:
Is this really doing what the docs say? I'm quite confused by this method. What is it extracting?